lib/rules/no-constant-condition.js

```diff
@@ -138,15 +139,28 @@ module.exports = {
                 case "Literal":
                 case "ArrowFunctionExpression":
                 case "FunctionExpression":
-                case "ObjectExpression":
                 case "ClassExpression":
                     return true;
+                case "ObjectExpression":
+                    if (!inBooleanPosition) {
+                        return node.properties.every(prop => {
+                            if (prop.type === "SpreadElement") {
+                                return isConstant(prop, false);
+                            }
+                            if (prop.type === "Property") {
+                                return prop.key.name !== "toString" && prop.key.name !== "valueOf";
```

Comment:
Oops. I think I failed to account for other types of keys here, like `{1: foo}` and `{[someExpression]: foo}`.

I'll update the PR to fix that.

Reply:
Actually, before I dig into this too much more (it's a little complicated), we should decide: Do we care about false positives where users are redefining `toString` or `valueOf`? If not, I can skip all of this.

If we do, we'll need to ensure we catch things like `{["toString"]: () => foo}` as well as `{[`toString`]: () => foo}` etc.

@nzakas

Reply:
I think we can skip the toString and valueOf cases, at least for now. Those seem like extremely rare cases so I’m fine with false positives on those.

Reply:
I've updated the PR.